src/com/amazon/ion/impl/IonCursorBinary.java

```diff
@@ -178,8 +262,362 @@ class IonCursorBinary implements IonCursor {
+            }
+        }
+        validate(configuration);
+        peekIndex = 0;
+        checkpoint = 0;
+
+        for (int i = 0; i < CONTAINER_STACK_INITIAL_CAPACITY; i++) {
+            containerStack[i] = new Marker(-1, -1);
+        }
+
+        this.buffer = new byte[configuration.getInitialBufferSize()];
+        this.startOffset = 0;
+        this.offset = 0;
+        this.limit = 0;
+        if (alreadyReadLen > 0) {
+            System.arraycopy(alreadyRead, alreadyReadOff, buffer, 0, alreadyReadLen);
+            limit = alreadyReadLen;
+        }
+        byteBuffer = ByteBuffer.wrap(buffer, 0, configuration.getInitialBufferSize());
+        refillableState = new RefillableState(
+            inputStream,
+            configuration.getInitialBufferSize(),
+            configuration.getMaximumBufferSize()
+        );
+    }
+
+    /* ---- Begin: internal buffer manipulation methods ---- */
+
+    /**
+     * @param index a byte index in the buffer.
+     * @return the number of bytes available in the buffer after the given index.
+     */
+    private long availableAt(long index) {
+        return limit - index;
+    }
+
+    /**
+     * Ensures that there is space for at least 'minimumNumberOfBytesRequired' additional bytes in the buffer,
+     * growing the buffer if necessary. May consolidate buffered bytes to the beginning of the buffer, shifting indices
+     * accordingly.
+     * @param minimumNumberOfBytesRequired the minimum number of additional bytes to buffer.
+     * @return true if the buffer has sufficient capacity; otherwise, false.
+     */
+    private boolean ensureCapacity(long minimumNumberOfBytesRequired) {
+        if (freeSpaceAt(offset) >= minimumNumberOfBytesRequired) {
+            // No need to shift any bytes or grow the buffer.
+            return true;
+        }
+        int maximumFreeSpace = refillableState.maximumBufferSize;
+        int startOffset = (int) offset;
+        if (minimumNumberOfBytesRequired > maximumFreeSpace) {
+            refillableState.isSkippingCurrentValue = true;
+            return false;
+        }
+        long shortfall = minimumNumberOfBytesRequired - refillableState.capacity;
+        if (shortfall > 0) {
+            int newSize = (int) Math.min(Math.max(refillableState.capacity * 2, nextPowerOfTwo((int) (refillableState.capacity + shortfall))), maximumFreeSpace);
+            byte[] newBuffer = new byte[newSize];
+            moveBytesToStartOfBuffer(newBuffer, startOffset);
+            refillableState.capacity = newSize;
+            buffer = newBuffer;
+            byteBuffer = ByteBuffer.wrap(buffer, (int) offset, (int) refillableState.capacity);
+        } else {
+            // The current capacity can accommodate the requested size; move the existing bytes to the beginning
+            // to make room for the remaining requested bytes to be filled at the end.
+            moveBytesToStartOfBuffer(buffer, startOffset);
```

Comment:
Did you consider implementing this as a circular buffer that reused the space ahead of the current position instead of moving bytes around?

Reply:
Yes, I initially implemented the original incremental reader that way. I moved away from it because I found a lot of the buffering logic to be more complicated, particularly around determining how much unread data is left in the buffer. We mitigate the cost of the shift by only doing it when a value to be filled would cross the buffer boundary (not on each value to be filled), so the frequency of shift is inversely related to the size of the buffer. The number of bytes to shift is directly related to the size of the value that would cross the boundary.